### ISSUE TITLE
char-cnn model in RStudio

### ISSUE THREAD

**Issue by Bunyodjon** [rstudio/keras#695]
Dear Friends, 

I am very new to Python. I was reading this article https://arxiv.org/abs/1508.06615 which is about Character-Aware Neural Language Models by Yong Kim.  I found an adapted code https://nbviewer.jupyter.org/github/BrambleXu/nlp-beginner-guide-keras/blob/master/char-level-cnn/notebooks/char-cnn-zhang-with-keras-pipeline.ipynb  for a classification problem using Keras in Python. I am trying to implement this in RStudio but I couldn't find any resource to implement character level - classficiation problem in RStudio using Keras.  How can I implement this in RStudio: 

`# write all code in one cell 

#========================Load data=========================
import numpy as np
import pandas as pd

train_data_source = '../data/ag_news_csv/train.csv'
test_data_source = '../data/ag_news_csv/test.csv'

train_df = pd.read_csv(train_data_source, header=None)
test_df = pd.read_csv(test_data_source, header=None)

#concatenate column 1 and column 2 as one text
for df in [train_df, test_df]:
    df[1] = df[1] + df[2]
    df = df.drop([2], axis=1)
    
#convert string to lower case 
train_texts = train_df[1].values 
train_texts = [s.lower() for s in train_texts] 

test_texts = test_df[1].values 
test_texts = [s.lower() for s in test_texts] 

#=======================Convert string to index================
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences

#Tokenizer
tk = Tokenizer(num_words=None, char_level=True, oov_token='UNK')
tk.fit_on_texts(train_texts)
#If we already have a character list, then replace the tk.word_index
#If not, just skip below part

#-----------------------Skip part start--------------------------
#construct a new vocabulary 
alphabet="abcdefghijklmnopqrstuvwxyz0123456789 ,;.!?:'\"/\\|_@#$%^&*~`+-=<>()[]{}"
char_dict = {}
for i, char in enumerate(alphabet):
    char_dict[char] = i + 1
    
#Use char_dict to replace the tk.word_index
tk.word_index = char_dict.copy() 
#Add 'UNK' to the vocabulary 
tk.word_index[tk.oov_token] = max(char_dict.values()) + 1
#-----------------------Skip part end----------------------------

#Convert string to index 
train_sequences = tk.texts_to_sequences(train_texts)
test_texts = tk.texts_to_sequences(test_texts)

#Padding
train_data = pad_sequences(train_sequences, maxlen=1014, padding='post')
test_data = pad_sequences(test_texts, maxlen=1014, padding='post')

#Convert to numpy array
train_data = np.array(train_data, dtype='float32')
test_data = np.array(test_data, dtype='float32')

#=======================Get classes================
train_classes = train_df[0].values
train_class_list = [x-1 for x in train_classes]

test_classes = test_df[0].values
test_class_list = [x-1 for x in test_classes]

from keras.utils import to_categorical
train_classes = to_categorical(train_class_list)
test_classes = to_categorical(test_class_list)`
`print(tk.word_index)`
`vocab_size=len(tk.word_index)
vocab_size
embedding_weights=[]
embedding_weights.append(np.zeros(vocab_size))
for char, i in tk.word_index.items(): # from index 1 to 70
    onehot = np.zeros(vocab_size)
    onehot[i-1] = 1
    embedding_weights.append(onehot)
embedding_weights = np.array(embedding_weights)
print(embedding_weights.shape) # first row all 0 for PAD, 69 char, last row for UNK
embedding_weights
from keras.layers import Input, Embedding, Activation, Flatten, Dense
from keras.layers import Conv1D, MaxPooling1D, Dropout
from keras.models import Model

#parameter 
input_size = 1014

#vocab_size = 69
embedding_size = 69
conv_layers = [[256, 7, 3], 
               [256, 7, 3], 
               [256, 3, -1], 
               [256, 3, -1], 
               [256, 3, -1], 
               [256, 3, 3]]

fully_connected_layers = [1024, 1024]
num_of_classes = 4
dropout_p = 0.5
optimizer = 'adam'
loss = 'categorical_crossentropy'
#Embedding layer Initialization
embedding_layer = Embedding(vocab_size+1, 
                            embedding_size,
                            input_length=input_size,
                            weights=[embedding_weights])
#Model 

#Input
inputs = Input(shape=(input_size,), name='input', dtype='int64')  # shape=(?, 1014)
#Embedding 
x = embedding_layer(inputs)
#Conv 
for filter_num, filter_size, pooling_size in conv_layers:
    x = Conv1D(filter_num, filter_size)(x) 
    x = Activation('relu')(x)
    if pooling_size != -1:
        x = MaxPooling1D(pool_size=pooling_size)(x) # Final shape=(None, 34, 256)
x = Flatten()(x) # (None, 8704)
#Fully connected layers 
for dense_size in fully_connected_layers:
    x = Dense(dense_size, activation='relu')(x) # dense_size == 1024
    x = Dropout(dropout_p)(x)
#Output Layer
predictions = Dense(num_of_classes, activation='softmax')(x)
#Build model
model = Model(inputs=inputs, outputs=predictions)
model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy']) # Adam, categorical_crossentropy
model.summary()

#1000 training samples and 100 testing samples
indices = np.arange(train_data.shape[0])
np.random.shuffle(indices)

x_train = train_data[indices][:1000]
y_train = train_classes[indices][:1000]

x_test = test_data[:100]
y_test = test_classes[:100]
#training
model.fit(x_train, y_train,
          validation_data=(x_test, y_test),
          batch_size=128,
          epochs=10,
          verbose=2)`

This code runs great. I would really apprecitate if someone can help me with building the layers for the model.  I can preprocess the data myself. I am have very poor understanding of building layers.

**Comment by skeydan** [comment]
Hi, normally you can pretty well translate the Python Keras code to its R equivalent.
To get started with the R syntax for building models, I suggest taking a look at the tutorials on

https://tensorflow.rstudio.com/keras/,

e.g. this: https://tensorflow.rstudio.com/keras/articles/tutorial_basic_text_classification.html

Then, you will find a lot of example code here:

https://tensorflow.rstudio.com/keras/articles/examples/
